### ISSUE TITLE
differences between `-N0` and `--secondary=no` for transcript mapping

### ISSUE THREAD

**Issue by Magdoll** [lh3/minimap2#98]
Hi,

I notice that if I run `--secondary=no`, there would possibly be multi-mappings, as shown by the matchAnnot report below where it shows **X were multiply mapped**.

Running `minimap2 -ax splice -uf -N0`
```
summary:   11822 isoforms read
summary:   11822 isoforms aligned, of which 0 were multiply mapped
summary:   11469 isoforms hit at least one gene, of which 89 were on opposite strand

summary:    8642 isoforms scored 5, of which   2557 had splice termination motif
summary:       0 isoforms scored 4, of which      0 had splice termination motif
summary:     676 isoforms scored 3, of which    124 had splice termination motif
summary:    1952 isoforms scored 2, of which    245 had splice termination motif
summary:       0 isoforms scored 1, of which      0 had splice termination motif
summary:     199 isoforms scored 0, of which      8 had splice termination motif
```

Running `minimap2 -ax splice -uf -secondary=no`
```
summary:   21259 isoforms read
summary:   21259 isoforms aligned, of which 7812 were multiply mapped
summary:   15651 isoforms hit at least one gene, of which 271 were on opposite strand

summary:   10839 isoforms scored 5, of which   2812 had splice termination motif
summary:       0 isoforms scored 4, of which      0 had splice termination motif
summary:     823 isoforms scored 3, of which    131 had splice termination motif
summary:    3116 isoforms scored 2, of which    278 had splice termination motif
summary:       0 isoforms scored 1, of which      0 had splice termination motif
summary:     873 isoforms scored 0, of which     14 had splice termination motif
```

In the `--secondary=no` instance, there were a lot more multimappings (7812 were multiply mapped). 

What does this mean? Does it mean that the 7812 were chimerically mapped (one sequence broken in multiple pieces to map to different loci) or multimapped?

**Comment by lh3** [comment]
> Running minimap2 -ax splice -uf -secondary=no

Are you using `--secondary=no` or `-secondary=no` (two or one hyphens)? Which minimap2 version are you using? With `--secondary=no` on your sample fastq, I see no reads mapped with SAM flag 0x100 (secondary flag). The output SAM file size is nearly the same.

**Comment by Magdoll** [comment]
I'm using `--secondary=no` (two hyphens).

Here is the command used which includes version:

```
@PG     ID:minimap2     PN:minimap2     VN:2.1.1-r341   CL:minimap2 -t 12 -ax splice -uf --secondary=no /home/UNIXHOME/etseng/share/minimap2_db/hg38/hg38.mmi touse.rep.fq
```

I'm including the  FASTQ file so you can try to replicate results:
https://www.dropbox.com/s/65jxdbbcw5n2634/touse.rep.fq.gz?dl=0

**Comment by lh3** [comment]
With `--secondary=no`, I get a SAM file with 13642 lines in total. With `-N0`, the output SAM has 13646 lines, very close. `--secondary=no` doesn't lead to obviously more chimeric alignments. This option should not generate any alignments with SAM flag 0x100 (secondary).

What I don't understand is why with `-N0`, you have "11822 isoforms read" but with `--secondary=no`, you have "21259 isoforms read". The difference between `-N0` and `--secondary=no` should be very small (with `--secondary=no` generally better).

What minimap2 version are you using? What is the file size of the outputted SAM with `--secondary=no` and with `-N0`?

**Comment by Magdoll** [comment]
I'm using minimap2 version 2.1.1-r341. I also found I am actually using `-secondary=no` (one dash). If I use two dashes, minimap2 complains this parameter does not exist.
```
bash-4.1$ minimap2 --version
2.1.1-r341
```

I have uploaded all the data to dropbox and the commands are below:
https://www.dropbox.com/sh/cgjhx6knht4kw4o/AAAQ_MuYMGv5rUKp_z1UH2zQa?dl=0

```
minimap2 -t 30 -ax splice -uf -N0 hg38.mmi touse.rep.fq > touse.rep.fq.N0.sam

minimap2 -t 30 -ax splice -uf -secondary=no hg38.mmi touse.rep.fq > touse.rep.fq.secondary_no.sam
```

Here's the SAM record difference:
```
bash-4.1$ wc -l *sam
    13907 touse.rep.fq.N0.sam
    21716 touse.rep.fq.secondary_no.sam
```

Then after running matchAnnot, you see the difference in number of records reflected as well.
--Liz

**Comment by lh3** [comment]
Could you try the latest version? v2.1.1 does not have --secondary=no. Ouch! I will throw an error when there is an unknown option. The system getopt_long() gives a warning, but the one I am using is not. Sorry!

**Comment by Magdoll** [comment]
Hi I have upgraded to release 2.7-r654 and now I can use the double dash `--secondary=no`. No secondary alignments found this time. 

I've updated my associated wiki page: https://github.com/Magdoll/cDNA_Cupcake/wiki/Best-practice-for-aligning-Iso-Seq-to-reference-genome:-minimap2,-GMAP,-STAR,-BLAT and updated both the dropbox link and the stats that compared minimap2 to GMAP.

I did notice that for the Lexogen SIRV synthetic spike in, there was a difference in recovered transcripts after switching to this new release with `--secondary=no`. I'll follow this case up in a different issue filing when I have more details.
